forward/forward.go

```diff
@@ -66,3 +68,79 @@ func Wrap(msg stanza.Message, body string, received time.Time, r xml.TokenReader
 		}.Wrap(r),
 	))
 }
+
+type forwardUnwrapper struct {
+	r         xml.TokenReader
+	stanzaXML xml.TokenReader
+	del       *delay.Delay
+}
+
+// Token implements xml.TokenReader
+func (f *forwardUnwrapper) Token() (xml.Token, error) {
+	if f.stanzaXML != nil {
+		token, err := f.stanzaXML.Token()
+		if err != nil && err != io.EOF {
+			// something bad happened
+			return nil, err
+		}
+		if err == io.EOF {
+			// we have reached the end of the stanza xml stream
+			f.stanzaXML = nil
+		}
+		if token != nil {
+			return token, nil
+		}
+	}
+
+	for {
```

Comment:
I think we can get rid of `forwardUnwrapper.innerXML` and this whole thing can be done in one pass without the looping. Instead we'd pop the token, check if it's a start element and a delay element at level 1 and if so decode it (or skip it if necessary) and return the next token, or just return the original token otherwise.

We'd also have to keep track of the level and whether or not a delay has already been found (I should have written a test for this, but if multiple are found I'm going to guess we should just use the first one).

Reply:
I attempted to implement this in my last change. 
In case there are multiple delay elements, I assumed we'd only consider the first one (to unmarshal/skip) and the others will be included in the returned stream. I've also added some test cases.

Why would we assume receiving multiple delay elements though? Should not we rather return an error in this case?

Reply:
Looks good to me. I could see us returning an error if we received multiple delay elements too, but I could go either way. It's probably safer to error if we get multiple at the top level, but I'm not sure what the user would be expected to do about them either.